python/sdist/amici/sbml_import.py

```diff
@@ -1027,6 +1006,137 @@ def _process_rules(self) -> None:
                                 'value')
             )
 
+    def _process_rule_algebraic(self, rule: sbml.AlgebraicRule):
+        formula = self._sympy_from_sbml_math(rule)
+        if formula is None:
+            return
+
+        free_variables = set()
+        # SBML L3V2 spec, p. 61:
+        # "Therefore, if an algebraic rule is introduced in a model,
+        # for at least one of the entities referenced in the rule’s
+        # math element the value of that entity must not be
+        # completely determined by other constructs in the model"
+        # find those elements:
+        for symbol in formula.free_symbols:
+            sbml_var = self.sbml.getElementBySId(str(symbol))
+            # This means that at least this entity must
+            # not have the attribute constant=“true”
+            if sbml_var.isSetConstant() and sbml_var.getConstant():
+                continue
+            # and there must also not be a rate rule or assignment
+            # rule for it
+            if self.is_assignment_rule_target(sbml_var) or \
+                    self.is_rate_rule_target(sbml_var):
+                continue
+            # Furthermore, if the entity is a Species object, its value
+            # must not be determined by reactions, which means that it
+            # must either have the attribute boundaryCondition=“false”
+            # or else not be involved in any reaction at all.
+            is_species = isinstance(sbml_var, sbml.Species)
+            is_boundary_condition = is_species and \
+                sbml_var.isSetBoundaryCondition() and \
+                sbml_var.getBoundaryCondition()
+            is_involved_in_reaction = is_species and \
+                not smart_is_zero_matrix(self.stoichiometric_matrix[
+                    list(self.symbols[SymbolId.SPECIES].keys()).index(symbol),
+                    :
+                ])
+            if is_species and not is_boundary_condition and is_involved_in_reaction:
+                continue
+            free_variables.add(symbol)
+
+        assert len(free_variables) >= 1
```

Comment:
Is that supposed to be caught by SBML validation? If so, good. If not, a more informative message would be nice.